### PR TITLE
Generate Serialization for WebCore::Font and WebCore::FontCustomPlatformData

### PR DESCRIPTION
**Pull request by sheeparegreat**
#### 05129ff5e069ff1d09d103b9de6878124a0c89e7
<pre>
Generate Serialization for WebCore::Font and WebCore::FontCustomPlatformData
<a href="https://bugs.webkit.org/show_bug.cgi?id=268527">https://bugs.webkit.org/show_bug.cgi?id=268527</a>
<a href="https://rdar.apple.com/122063976">rdar://122063976</a>

Reviewed by Alex Christensen.

* Source/WebCore/platform/graphics/Font.cpp:
(WebCore::Font::create):
* Source/WebCore/platform/graphics/Font.h:
* Source/WebCore/platform/graphics/FontCustomPlatformData.h:
* Source/WebCore/platform/graphics/FontPlatformData.cpp:
* Source/WebCore/platform/graphics/FontPlatformData.h:
* Source/WebCore/platform/graphics/coretext/FontCustomPlatformDataCoreText.cpp:
(WebCore::FontCustomPlatformData::tryMakeFromSerializationData):
(WebCore::FontCustomPlatformData::serializedData const):
* Source/WebCore/platform/graphics/coretext/FontPlatformDataCoreText.cpp:
(WebCore::FontPlatformData::tryMakeFontPlatformData):
(WebCore::FontPlatformData::FontPlatformData):
(WebCore::FontPlatformData::platformSerializationData const):
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.mm:
(IPC::ArgumentCoder&lt;WebCore::Font&gt;::encodePlatformData): Deleted.
(IPC::ArgumentCoder&lt;WebCore::Font&gt;::decodePlatformData): Deleted.
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder&lt;WebCore::Font&gt;::encode):
(IPC::ArgumentCoder&lt;Font&gt;::decode):
(IPC::ArgumentCoder&lt;FontPlatformData::Attributes&gt;::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/274190@main">https://commits.webkit.org/274190@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/9f40a5c0749c0be27ba820e36811ec66b2ea49c1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/38169 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/17089 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/40463 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/40718 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/33957 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/40415 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/19813 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/14430 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/32236 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/38741 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/14422 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/33392 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12570 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/12534 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/34137 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/42007 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/34658 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/34580 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/38396 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/13126 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/10800 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36592 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/14665 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/13542 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4970 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/14116 "Built successfully") | | |

### CODE REVIEW

File: Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in [hunk 1]
```diff
@@ -22,6 +22,44 @@
...
 
+#if USE(CORE_TEXT)
+class WebCore::FontPlatformData {
+    float size();
+    WebCore::FontOrientation orientation();
+    WebCore::FontWidthVariant widthVariant();
+    WebCore::TextRenderingMode textRenderingMode();
+    bool syntheticBold();
+    bool syntheticOblique();
+    WebCore::FontPlatformData::PlatformDataVariant platformSerializationData();
```
Comment (achristensen07) [inline Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in]: This should be std::variant<FontPlatformDataSerialization, FontPlatformDataSerializationWithCreationData> or there should be a using statement to put into the metadata what this is.

File: Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in [hunk 2]
```diff
@@ -22,6 +22,44 @@
...
+}
+
+header: <WebCore/FontPlatformData.h>
+[CustomHeader] struct WebCore::FontPlatformDataSerializationWithCreationData {
+    Vector<uint8_t> fontFaceData;
+    RetainPtr<CFDictionaryRef> attributes;
```
Comment (achristensen07) [inline Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:39]: This used to always be encoded, and now it is encoded only if the PlatformDataVariant is FontPlatformDataSerializationWithCreationData.  Is that an intentional change?

File: Source/WebCore/platform/graphics/FontPlatformData.cpp [hunk 3]
```diff
@@ -114,4 +115,68 @@ Vector<FontPlatformData::FontVariationAxis> FontPlatformData::variationAxes(Shou
 }
 #endif
 
+#if USE(CORE_TEXT)
```
Comment (donny-dont) [inline Source/WebCore/platform/graphics/FontPlatformData.cpp]: Might it be better to move this into its `Source/WebCore/platform/graphics/coretext/FontPlatformDataCoreText.cpp`?

File: Source/WebCore/platform/graphics/FontPlatformData.cpp [hunk 4]
```diff
@@ -114,4 +115,68 @@ Vector<FontPlatformData::FontVariationAxis> FontPlatformData::variationAxes(Shou
...
+bool FontPlatformData::canMakeBaseFontDescriptor(const Vector<uint8_t>& fontFaceData, String& itemInCollection)
+{
+    Vector<uint8_t> copyOfFontFaceData = fontFaceData;
+    auto localFontFaceData = SharedBuffer::create(WTFMove(copyOfFontFaceData));
+    auto fontCustomPlatformData = createFontCustomPlatformData(localFontFaceData, itemInCollection);
```
Comment (achristensen07) [inline Source/WebCore/platform/graphics/FontPlatformData.cpp]: This is a lot of unnecessary copying and duplicate work just for a verifier.  Could we instead just try making the object and validate that it succeeded?

File: Source/WebCore/platform/graphics/FontPlatformData.h [hunk 5]
```diff
@@ -133,6 +133,21 @@ struct FontPlatformDataAttributes {
 #endif
 };
 
+#if USE(CORE_TEXT)
+struct FontPlatformDataSerializationWithCreationData {
```
Comment (achristensen07) [inline Source/WebCore/platform/graphics/FontPlatformData.h]: Name is a mouthful.  SerializedCreationData?

File: Source/WebCore/platform/graphics/FontPlatformData.cpp [hunk 6]
```diff
@@ -114,4 +115,68 @@ Vector<FontPlatformData::FontVariationAxis> FontPlatformData::variationAxes(Shou
...
+    if (!fontCustomPlatformData)
+        return false;
+    auto baseFontDescriptor = fontCustomPlatformData->fontDescriptor.get();
+    if (!baseFontDescriptor)
+        return false;
+    return true;
+}
```
Comment (donny-dont) [inline Source/WebCore/platform/graphics/FontPlatformData.cpp]: Seems kinda wasteful to go through the creation of a `FontCustomPlatformData` and not use it. Not sure what the better way is here just making a note.

**Review by achristensen07 — CHANGES_REQUESTED** [review]

**Review by donny-dont — COMMENTED** [review]
Just had a few thoughts. Will let @cdumez and @achristensen07 chime in for the specifics of the platform.

Will work on the Windows and FreeType variant once this lands. Thanks for getting this started!

File: Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in (hunk 2 above)
Comment (sheeparegreat) [reply Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:39]: `RetainPtr<CFDictionaryRef> attributes` is in both structs.

File: Source/WebCore/platform/graphics/FontCustomPlatformData.h [hunk 7]
```diff
@@ -74,6 +82,10 @@ struct FontCustomPlatformData : public RefCounted<FontCustomPlatformData> {
 
     FontPlatformData fontPlatformData(const FontDescription&, bool bold, bool italic, const FontCreationContext&);
 
+#if USE(CORE_TEXT)
+    WEBCORE_EXPORT FontCustomPlatformSerializedData SerializedData() const;
```
Comment (achristensen07) [inline Source/WebCore/platform/graphics/FontCustomPlatformData.h]: s should be lower case.

File: Source/WebCore/platform/graphics/FontPlatformData.h [hunk 8]
```diff
@@ -133,6 +133,21 @@ struct FontPlatformDataAttributes {
 #endif
 };
 
+#if USE(CORE_TEXT)
+struct FontPlatformSerializedCreationData {
+    Vector<uint8_t> fontFaceData { };
```
Comment (achristensen07) [inline Source/WebCore/platform/graphics/FontPlatformData.h]: { } is unnecessary

File: Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in [hunk 9]
```diff
@@ -22,6 +22,46 @@
...
+using WebCore::FontPlatformData::PlatformDataVariant = std::variant<FontPlatformSerializedData, FontPlatformSerializedCreationData>;
+
+[CreateUsing=tryMakeFontPlatformData] class WebCore::FontPlatformData {
+    float size();
+    WebCore::FontOrientation orientation();
+    WebCore::FontWidthVariant widthVariant();
+    WebCore::TextRenderingMode textRenderingMode();
+    bool syntheticBold();
+    bool syntheticOblique();
+    WebCore::FontPlatformData::PlatformDataVariant platformSerializationData();
```
Comment (achristensen07) [inline Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in]: "std::variant" should be in this line in the serialization.in file.  Even though the types are the same, the metadata needs to be more explicit.

**Review by achristensen07 — CHANGES_REQUESTED** [review]

**Review by achristensen07 — APPROVED** [review]

**Pull request merged**